papers/TIARA/TIARA/src/retriever/freebase_cache_utils.py

```diff
@@ -0,0 +1,152 @@
+                        neighbors = retriever.neighbor_value_by_mid_and_relation(ns(mid), ns(r),
+                                                                                 forward=True,
+                                                                                 backward=False)
+                        if len(neighbors) != 0:
+                            forward_neighbors[mid][r] = neighbors
+                    except Exception as e:
+                        print(e)
+            if count % 50 == 0:
+                print('[INFO] forward neighbors: ' + str(count) + ' / ' + str(len(retriever.forward_relations)))
+            count += 1
+        print('\n')
+
+    count = 0
+    if retriever.backward_relations is not None:
+        for mid in retriever.backward_relations.keys():
+            if mid not in backward_neighbors:
+                backward_neighbors[mid] = dict()
+            for r in retriever.backward_relations[mid]:
+                r = remove_ns(r)
+                if r not in backward_neighbors[mid]:
+                    try:
+                        neighbors = retriever.neighbor_value_by_mid_and_relation(ns(mid), ns(r),
+                                                                                 forward=False,
+                                                                                 backward=True)
+                        if len(neighbors) != 0:
+                            backward_neighbors[mid][r] = neighbors
+                    except Exception as e:
+                        print(e)
+            if count % 50 == 0:
+                print('[INFO] backward neighbors: ' + str(count) + ' / ' + str(len(retriever.backward_relations)))
+            count += 1
+        print('\n')
+
+    pickle_save(forward_neighbors, output_dir + '/forward_nei.bin')
+    pickle_save(backward_neighbors, output_dir + '/backward_nei.bin')
+
+
+def create_rdf_type_cache():
+    retriever = FreebaseRetriever()
+    res = dict()
+    count = 0
+
+    file_path_list = ['../dataset/GrailQA/grailqa_v1.0_train.json',
+                      '../dataset/GrailQA/grailqa_v1.0_dev.json',
+                      '../dataset/GrailQA/grailqa_el.json']
+    for file_path in file_path_list:
+        count = 0
+        uri_set = None
+        if 'grailqa_v' in file_path:
+            dataloader = GrailQAJsonLoader(file_path)
+            uri_set = dataloader.get_uris()
+        elif 'grailqa_el' in file_path:
+            linker = GrailQAEntityLinker(file_path)
+            uri_set = linker.get_mid_set()
+
+        for uri in uri_set:
+            if uri not in res:
+                res[uri] = retriever.rdf_type_by_uri(uri)
+
+            if count % 100 == 0:
+                print('[INFO] ' + file_path + ' rdf type: ' + str(count) + ' / ' + str(len(uri_set)))
+            count += 1
+
+    pickle_save(res, freebase_cache_dir + '/rdf_type.bin')
+
+
+if __name__ == "__main__":
+    # create_relation_cache()
+    # create_neighbor_cache()
```

Comment:
We'll provide all pre-gen caches in the data download, right?

Reply:
All the caches can be generated by our scripts